### ISSUE TITLE
Question: On global behavior sideeffects

### ISSUE THREAD

**Issue by aligthart** [mailgun/gubernator#44]
See https://github.com/mailgun/gubernator/blob/master/architecture.md#side-effects-of-global-behavior

It states:  _Using GLOBAL can increase the amount of traffic per rate limit request if the cluster is large enough_

I assume that with '_the cluster_' you mean the cluster of gubernator instances.

Can you give a more detailed indication what you mean with '_large'? Are we talking about tens or hundreds of gubernator instances. Is is still practical to run 20 instances of gubernator in global mode. Or would that already kill the network. Is there a practical limit to the number of gubernator instances running in global mode?

Would be nice if that could be explained in the docs.

**Comment by thrawn01** [comment]
I meant that documentation only as a note to anyone using global but not as a warning against using it. 

Global requests can be forwarded asynchronously to the owning peer, then the owning peer will update every node in the cluster with the rate limit status. So assuming a single node cluster, for each request that comes in, there is a minimum of 2 additional (3 if the receiving node isn't the owner) network updates before all nodes have the `Hit` updated in their cache. 

To calculate the WORST case scenario, we total the number of network updates that must occur for each global rate limit. 
* Count 1 incoming request to the node
* Count 1 request when forwarding to the owning node
* Count 1 + (number of nodes in cluster) to update all the nodes with the current `Hit` count.

Remember this is the WORST case, as the node that recieved the request might be the owning node thus no need to forward to the owner. Additionally we improve the worst case by having the owning node batch `Hits` when forwarding to all the nodes in the cluster. Such that a 1,000 individual requests of`Hit = 1` each for a unique key will result in batched request from the owner to each node with a single `Hit = 1,000` update. 

Additionally thousands of hits to different unique keys will also be batched such that network usage doesn't increase until the number of requests in an update batch exceeds the `BehaviorConfig.GlobalBatchLimit` or when the number of nodes in the cluster increases. (thus more batch updates) When that occurs you could increase the `GlobalBatchLimit` (defaults to 1,000) to increase the number of unique key `Hit` updates that are batched into a single update request. This will lower the number of update requests made to each node in the cluster, thus increase network efficiency.

I personally would not expect to see network contention until the cluster reaches hundreds of nodes and fielding millions of **unique key** requests per second. 

NOTE: Requests for the **same** key is super efficient, the worry here is **unique key** requests, as this increases the number of updates batched.

Gubernator is every efficient when using GLOBAL behavior, The documentation note is only there to inform users of the additional network communication when using GLOBAL. More communication is one of the trade offs for lower latency and higher throughput you recieve when using GLOBAL. This could be a major implication if you are trying to run gubernator cluster across multiple regions. (I'm working on proper cross region support now)

TLDR; It can be more efficient to use GLOBAL or more efficient to use the standard behavior depending on the use case. Does this make sense? Help in documenting this would be much appreciated if you have an idea of how to document this clearly.

**Comment by aligthart** [comment]
Thanks for the explanation. Its clear now.
Maybe what could help in the documentation is adding a sequence diagram for each type of behavior describing how requests are handled. Up to you guys but feel free to close this ticket.

As a note. Maybe another benefit of using GLOBAL behavior is the minimal amount of moving parts. We are looking at running the Gubernater container as a sidecar to nginx so that nginx can use 'localhost' to talk to its local Gubernator container (so no external network traffic). When running with GLOBAL behavior Gubernator can server the request from local cache (again no network traffic). As a result the changes of something going wrong with throttling requests (or not being able to handle requests at all due to throttling issues) is reduced to a minimum (important to us). We then rely on the asynchronous GLOBAL behavior for cache updates and if something would go wrong there, it would not negatively impact throttled tenants. Also scaling up/down and restarting Gubernator containers will now have minimal side effects we think as this now runs as part of nginx.
